src/cockpit/remote.py

```diff
@@ -15,21 +15,185 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <https://www.gnu.org/licenses/>.
 
+import asyncio
+import getpass
+import logging
+import os.path
+import re
+import socket
 
-from typing import Dict
+from typing import Dict, Optional, Tuple
 
-from .router import RoutingError, RoutingRule
+from ._vendor import ferny
+from .router import RoutingRule, Router
+from .peer import Peer, PeerError
+
+logger = logging.getLogger(__name__)
+
+
+class PasswordResponder(ferny.InteractionResponder):
+    PASSPHRASE_RE = re.compile(r"Enter passphrase for key '(.*)': ")
+
+    password: Optional[str]
+    error_message: Optional[str]
+    attempts: int
+
+    def __init__(self, password: Optional[str]):
+        self.password = password
+        self.error_message = None
+        self.attempts = 0
+
+    async def do_hostkey(self, reason: str, host: str, algorithm: str, key: str, fingerprint: str) -> bool:
+        try:
+            logger.debug('checking ~/.ssh/known_hosts')
+            with open(os.path.expanduser('~/.ssh/known_hosts')) as file:
+                logger.debug('  file is open')
+                if key in file.read():
+                    logger.debug('  and contains the key!')
+                    return False  # this looks like it's going to succeed on its own
+                logger.debug('  but does not contain the key :(')
+        except OSError as exc:
+            logger.debug('  err: %s', exc)
+
+        raise PeerError('unknown-hostkey', error='unknown-hostkey', auth_method_results={},
+                        host_key=f'{host} {algorithm} {key}', host_fingerprint=fingerprint)
+
+    async def do_askpass(self, messages: str, prompt: str, hint: str) -> Optional[str]:
+        logger.debug('Got askpass(%s): %s', hint, prompt)
+
+        match = PasswordResponder.PASSPHRASE_RE.fullmatch(prompt)
+        if match is not None:
+            path = match.group(1)
+            logger.debug("This is a passphrase request for %s, but we don't do those.  Abort.", path)
+            self.error_message = f'locked identity: {path}'
+            return None
```

Comment:
Fair enough -- Cockpit's UI does support it, but of course it's fine to implement it in a follow-up.

Reply:
This is actually what's supposed to happen.  We don't have support for unlocking locked keys, and we don't have any way to receive the passphrases to do so.